pkg/node/local_node_store.go

```diff
@@ -187,40 +123,77 @@ func NewLocalNodeStore(params LocalNodeStoreParams) (*LocalNodeStore, error) {
 	return s, nil
 }
 
+// Observe changes to the local node state.
+func (s *LocalNodeStore) Observe(ctx context.Context, next func(LocalNode), complete func(error)) {
+	go func() {
+		// Wait until initialized before starting to observe.
+		_, initWatch := s.nodes.Initialized(s.db.ReadTxn())
+		select {
+		case <-initWatch:
+		case <-ctx.Done():
+			complete(ctx.Err())
+			return
+		}
+		stream.Map(
+			stream.Filter(
+				statedb.Observable(s.db, s.nodes),
+				// Only care about non-deleted local nodes. The local node is never deleted.
+				func(ev statedb.Change[*LocalNode]) bool {
+					return !ev.Deleted && ev.Object != nil && ev.Object.Local != nil
+				},
+			),
+			func(ev statedb.Change[*LocalNode]) LocalNode {
+				return *ev.Object
+			},
+		).Observe(ctx, next, complete)
+	}()
+}
+
 // Get retrieves the current local node. Use Get() only for inspecting the state,
 // e.g. in API handlers. Do not assume the value does not change over time.
 // Blocks until the store has been initialized.
 func (s *LocalNodeStore) Get(ctx context.Context) (LocalNode, error) {
+	_, initWatch := s.nodes.Initialized(s.db.ReadTxn())
 	select {
-	case <-s.hasValue:
-		s.getMu.RLock()
-		defer s.getMu.RUnlock()
-
-		if s.complete == nil {
-			// Return EOF when the LocalNodeStore is stopped, to preserve the
-			// same behavior of stream.First[LocalNode].
-			return LocalNode{}, io.EOF
-		}
-
-		return s.value, nil
-
+	case <-initWatch:
 	case <-ctx.Done():
 		return LocalNode{}, ctx.Err()
 	}
+
+	ln, _, found := s.nodes.Get(s.db.ReadTxn(), LocalNodeQuery)
+	if !found {
+		return LocalNode{}, errors.New("Local node not found")
+	}
+	return *ln, nil
 }
 
-// Update modifies the local node with a mutator. The updated value
-// is passed to observers. Calling LocalNodeStore.Get() from the
-// mutation function is forbidden, and would result in a deadlock.
+// Update modifies the local node with a mutator.
 func (s *LocalNodeStore) Update(update func(*LocalNode)) {
-	s.mu.Lock()
-	defer s.mu.Unlock()
-
-	s.getMu.Lock()
-	update(&s.value)
-	s.getMu.Unlock()
+	txn := s.db.WriteTxn(s.nodes)
+	defer txn.Commit()
+	ln, _, found := s.nodes.Get(txn, LocalNodeQuery)
+	if !found {
+		panic("BUG: No local node exists")
+	}
+	ln = ln.DeepCopy()
```

Comment:
Most of the callers do already perform some amount of copying to not mutate the existing state. It may be worth cleaning that up at this point, to prevent unnecessary work.

Reply:
I was thinking as next step to remove uses of `LocalNodeStore.Get` completely and then we'll naturally get rid of this.

Reply:
> Or did you mean changing the callers not to unnecessarily copy stuff?

This basically, as they were previously based on the assumption that they were given a shallow copy. Probably not a big deal though, given that as you mentioned updates should be relatively rare.